Offline Mode: Remove post.status usage from the publishing flows

kean opened the pull request:
This PR fixes the issue where changes to `post.status` for draft posts were automatically synced to the server. The app used to update `post.status` locally when changing "Publish Date" and post visibility settings.

- Update `PublishSettingsViewModel` to only change `post.dateCreated` (represents "Publish Date" in WordPress) and never `post.status`
- Remove the "Visibility" cell from Post Settings when editing drafts. It would've been prohibitively expensive to re-implement the same behavior as in Gutenberg when changing "Visibility" would trigger the publishing flow: it's one entry for them, and a new one for the app.
- Remove the "Status" cell fro Post Settings form the Editor – this just plan made no sense. I'll add a "Pending review" toggle just like in Gutenberg in the upcoming PRs.
- The publishing options set in the publishing sheets no longer saved in the DB (except for tags and categories which are)
- Add SwiftUI `PostVisibilityPicker` – I had to change it anyway and I had a prototype already available in my "design" repo. The new version closer represents the Gutenberg one.
- (Extra) Rework `PrepublishingViewController.show` to reuse the code between the Editor and the Post List (will be needed later; technically shouldn't be part of this PR)

## Tests

### Publish Date

**Test 1.1**

1. Create a new draft
2. Open Post Settings
3. Select Publish Date in the future
4. Save 
5. **Verify** that the primary button in the editor changed to "Schedule"
6. Tap "Bac" and then "Save Draft"
7. **Verify** that the post was saved as a draft and was NOT published
10. Reopen the post
11. **Verify** that the primary button still says "Schedule"
12. **Verify** that the Publish Date was saved
13. Schedule a post and verify that it was scheduled for a selected date

**Test 1.2**

1. Create a new draft
2. Open Post Settings
3. Select Publish Date in the future
4. Save 
5. **Verify** that the primary button in the editor changed to "Schedule"
6. Open Post Settings
7. Select Publish Date in the past or "Now"
5. **Verify** that the primary button in the editor changed to "Publish"

**Test 1.3** (Continuity)

1. Create a new draft and save it
2. Set the publish date to a future date from the web (anywhere)
3. Re-open the post for editing from the app
4. **Verify** that the primary button says "Schedule"
5. Open post settings
6. **Verify** that it displayed the date set by the web
7. Change the publish date from the app and save
8. **Verify** that the changes were reflected on the web

### Post Settings

**Test 2.1**

1. Create a new draft
2. Open Post Settings
3. **Verify** that the "Status" and "Visibility" options are no longer available

### Visibility Picker

**Test 3.1**

1. Create a new draft
2. Tap "Publish"
3. Tap "Visibility"
4. **Verify** that the new screen is displayed that closely matches Gutenberg 
5. Select "Private" visibility
6. **Verify**  that the visibility picker was dismissed and the selection changed
7. Confirm publishing
8. **Verify** that the post was published as a private post (check on wp.com)

**Test 3.2**

1. Create a new draft
2. Tap "Publish"
3. Tap "Visibility"
5. Select "Password Protected" visibility
6. **Verify** that the password field appeared and the screen shifted up to accommodate the keyboard
7. Tap "Cancel" and **verify** that the previous visibility was selected
8. Select "Password Protected" again
9. **Verify** that the "View Password" and "Clear Password" buttons work
10. Tap "Save"
11. **Verify**  that the visibility picker was dismissed and the selection changed
12. Confirm publishing
13. **Verify** that the post was published as a private post (check on wp.com)

## Regression Notes
1. Potential unintended areas of impact: Publishing flow
14. What I did to test those areas of impact (or what existing automated tests I relied on): manual
15. What automated tests I added (or what prevented me from doing so): n/a

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

kean commented on WordPress/Classes/ViewRelated/Post/Prepublishing/PrepublishingViewController.swift:
```diff
@@ -498,7 +570,7 @@ final class PrepublishingViewController: UIViewController, UITableViewDataSource
                 control.isEnabled = !isLoading
             case let cell as UITableViewCell:
-                isLoading ? cell.disable() : cell.enable()
+                cell.textLabel?.textColor = isLoading ? .secondaryLabel : .label
```
A small tweak so that the disclosure indicators don't disappear during loading, which didn't look great.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

kean commented on WordPress/Classes/Models/AbstractPost.h:
```diff
@@ -20,11 +20,6 @@ typedef NS_ENUM(NSUInteger, AbstractPostRemoteStatus) {
 @property (nonatomic, strong) Blog *blog;
-/**
- The dateModified field is used in tandem with date_created_gmt to determine if
- a draft post should be published immediately. A draft post will "publish immediately"
```
Nope.

kean commented on WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift:
```diff
@@ -46,6 +46,7 @@ protocol PublishingEditor where Self: UIViewController {
     /// Debouncer used to save the post locally with a delay
     var debouncer: Debouncer { get }
 
+    /// - warning: deprecated (kahu-offline-mode)
```
It was needed only for stories – will remove later.

kean commented on WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift:
```diff
@@ -315,52 +316,8 @@ extension PublishingEditor {
-            completion(.confirmed)
-        }
-        present(alertController, animated: true, completion: nil)
+        PrepublishingViewController.show(for: post, action: action, from: self, completion: completion)
```
This just moves the code around (no functional changes).

kean committed:
Remove post.status setter usage

momo-ozawa reviewed (commented):
@kean I found a critical bug via test 1.2, and some other issues/observations:

### Test 1.2: Saving draft actually publishes post

Setting the published date to a future date then back to "Now" actually published the post when you tap "Save draft"

https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/49b2642b-bcea-4257-8bd2-0877104468f1

### Checkmark design discrepancy
IMO, using the same checkmark asset as the Gutenberg visibility setting would look nicer

Gutenberg | Visibility sheet 
-- | --
![Simulator Screenshot - iPhone 15 Pro - 2024-04-03 at 12 23 53](https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/3fd3b944-e1cc-4b12-a212-f7bd4d2f5306) | ![Simulator Screenshot - iPhone 15 Pro - 2024-04-03 at 12 24 30](https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/54d2ff7d-8291-4ff1-a64e-8f0f8f7273f2)

### Two taps required to change publish date 

What do you think of directly opening the date picker if a user taps on the "Publish Date" row in the post setting screen? 

https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/6f041223-53a9-4972-b7b2-e958eed356ac

### Prepublishing publish date typo
Should be "Immediately" 

<img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/0ea7eed7-6c28-4edd-8a76-9e49aa5b9d4b" width=200>

kean committed:
Fix an issue with changing publish date and status

kean committed:
Fix typos

kean committed:
Update checkmark style

kean commented:
> Setting the published date to a future date then back to "Now" actually published the post when you tap "Save draft"

Good catch. Fixed!

> IMO, using the same checkmark asset as the Gutenberg visibility setting

Updated. I'll integrate the new picker in Gutenberg as well if time allows.

> What do you think of directly opening the date picker if a user taps on the "Publish Date" row in the post setting screen?

Yes, it's pretty bad. I think it might be worth doing separately in the scope of this PR: https://github.com/wordpress-mobile/WordPress-iOS/pull/22923. It should also clearly display the timezone (your blog timezone; not your current).

> Should be "Immediately"

Fixed
